### PR TITLE
Update the example_config.py

### LINKED ISSUE #331: Update ExampleConfig to be more like the ocean.js ConfigHelper
ocean.js offers the `ConfigHelper.getConfig` method which takes a network name, and returns a dictionary of config parameters for the corresponding network like chainId, provider URL, aquarius URL, etc.  

ocean.py offers the `ExampleConfig.get_config` method that returns a `Config` instance that is used by unit tests but wouldn't work for a real network. 

This issue is to update the ocean.py `ExampleConfig` to be more like the ocean.js `ConfigHelper`.

See for details: 
1. https://github.com/oceanprotocol/ocean.js/blob/main/src/utils/ConfigHelper.ts
2. https://github.com/oceanprotocol/ocean.py/blob/main/ocean_lib/example_config.py

### PR DESCRIPTION
**Pull request by mariacarmina**
Fixes #331  .

Changes proposed in this PR:

- updated `example_config.py` to support ganache, ropsten, rinkeby, mainnet, polygon-mainnet and BSC;
- implemented unit tests  for the new config structure for each network which do not ask about security keys (only `BSC` and `ganache`);
- added more fields in the config object such as `chain_id`, `store_interval` and `network_name`.

### CODE REVIEW

**Comment by mariacarmina** [comment]
Succeed #427. This is the new PR that resolves #331.

File: ocean_lib/example_config.py
Comment (calina-c) [inline ocean_lib/example_config.py:19]: can you please keep the constants in one part, and Config and config_defaults either at the beginning of these imports or at the end?

File: ocean_lib/example_config.py
Comment (calina-c) [inline ocean_lib/example_config.py:115]: no need for the else. If the Error is raised, the code won't continue. Remove the else and unindent its contents.

**Review by calina-c — CHANGES_REQUESTED** [review]

File: ocean_lib/example_config.py
Comment (mariacarmina) [reply ocean_lib/example_config.py:19]: Sure

File: ocean_lib/example_config.py
Comment (mariacarmina) [reply ocean_lib/example_config.py:115]: Ok, will refactor right now.

**Review by calina-c — DISMISSED** [review]

**Comment by calina-c** [comment]
@mariacarmina please double check readmes before merging. I'd also like it if @DMats had a look at this, since he originated the ticket and the approach was adapted in the meantime.

File: ocean_lib/example_config.py
Comment (DMats) [inline ocean_lib/example_config.py:120]: This validation check is performed by `get_config_helper_network` so this line is a duplicate. Please remove it.

File: pytest.ini
Comment (DMats) [inline pytest.ini:17]: Please add a newline at the end of this file. As a general rule I put new lines at the end of all files.

File: ocean_lib/test/test_example_config.py
Comment (DMats) [inline ocean_lib/test/test_example_config.py:1]: I'd like to see unit tests for rinkeby, ropsten, and mainnet as well. After you integrate Calina's changes to the `get_web3_connection_provider` from https://github.com/oceanprotocol/ocean.py/pull/421, this should be possible.

**Review by DMats — CHANGES_REQUESTED** [review]
This PR is missing the changes from https://github.com/oceanprotocol/ocean.py/pull/421.  Please cherry-pick https://github.com/oceanprotocol/ocean.py/commit/0fc8d9b3eb6ce8f9726b43a808020ea2097b510f from the main branch.

**Comment by mariacarmina** [comment]
> @mariacarmina please double check readmes before merging. I'd also like it if @DMats had a look at this, since he originated the ticket and the approach was adapted in the meantime.

I have tested the marketplace-flow README right now and it seems to work fine. I'll let here all the process.
<details>
<summary> Marketplace-flow README - Success </summary>
<p>
(venv) maria@maria-VirtualBox:~/Desktop/test3$ python </br>
Python 3.8.10 (default, Jun  2 2021, 10:49:15)  </br>
[GCC 9.4.0] on linux </br>
Type "help", "copyright", "credits" or "license" for more information. </br>
>>> from ocean_lib.config import Config </br>
>>> from ocean_lib.ocean.ocean import Ocean </br>
>>> config = Config('config.ini') </br>
>>> ocean = Ocean(config) </br>
>>> print(f"config.network_url = '{config.network_url}'") </br>
config.network_url = 'ganache' </br>
>>> print(f"config.metadata_cache_uri = '{config.metadata_cache_uri}'") </br>
config.metadata_cache_uri = 'http://localhost:5000' </br>
>>> print(f"config.provider_url = '{config.provider_url}'") </br>
config.provider_url = 'http://localhost:8030' </br>
>>> import os </br>
>>> from ocean_lib.web3_internal.wallet import Wallet </br>
>>> alice_wallet = Wallet(ocean.web3, private_key=os.getenv('TEST_PRIVATE_KEY1')) </br>
>>> print(f"alice_wallet.address = '{alice_wallet.address}'") </br>
alice_wallet.address = '0x66aB6D9362d4F35596279692F0251Db635165871' </br>
>>> from ocean_lib.ocean.mint_fake_ocean import mint_fake_OCEAN </br>
>>> mint_fake_OCEAN(config) </br>
>>> assert alice_wallet.web3.eth.get_balance(alice_wallet.address) > 0, "need ETH" </br>
>>> data_token = ocean.create_data_token('DataToken1', 'DT1', alice_wallet,  blob=ocean.config.metadata_cache_uri) </br>
>>> token_address = data_token.address </br>
>>> print(f"token_address = '{token_address}'") </br>
token_address = '0x08787cC8dB325A9b84F60E99529C8000b4ef913c' </br>
>>> date_created = "2019-12-28T10:55:11Z" </br>
>>> metadata =  { </br>
...     "main": { </br>
...         "type": "dataset", "name": "branin", "author": "Trent", </br>
...         "license": "CC0: Public Domain", "dateCreated": date_created, </br>
...         "files": [{"index": 0, "contentType": "text/text", </br>
...                "url": "https://raw.githubusercontent.com/trentmc/branin/master/branin.arff"}]} </br>
... } </br>
>>> service_attributes = { </br>
...         "main": { </br>
...             "name": "dataAssetAccessServiceAgreement", </br>
...             "creator": alice_wallet.address, </br>
...             "timeout": 3600 * 24, </br>
...             "datePublished": date_created, </br>
...             "cost": 1.0, </br>
...         } </br>
...     } </br>
>>> from ocean_lib.data_provider.data_service_provider import DataServiceProvider </br>
>>> from ocean_lib.common.agreements.service_factory import ServiceDescriptor </br>
>>> service_endpoint = DataServiceProvider.get_url(ocean.config) </br>
>>> download_service = ServiceDescriptor.access_service_descriptor(service_attributes, service_endpoint) </br>
>>> assert alice_wallet.web3.eth.get_balance(alice_wallet.address) > 0, "need ETH" </br>
>>> asset = ocean.assets.create( </br>
...   metadata, </br>
...   alice_wallet, </br>
...   service_descriptors=[download_service], </br>
...   data_token_address=token_address) </br>
signing message with nonce 80: 0x08787cC8dB325A9b84F60E99529C8000b4ef913c, account=0x66aB6D9362d4F35596279692F0251Db635165871 </br>
>>> assert token_address == asset.data_token_address </br>
>>> did = asset.did  # did contains the datatoken address </br>
>>> print(f"did = '{did}'") </br>
did = 'did:op:08787cC8dB325A9b84F60E99529C8000b4ef913c' </br>
>>> data_token.mint_tokens(alice_wallet.address, 100.0, alice_wallet) </br>
'0x792c38df0960f041c794f36333bd510b3d816dac6ec2f5412bf00c75b5c5ee3d' </br>
>>> from ocean_lib.models.btoken import BToken #BToken is ERC20 </br>
>>> OCEAN_token = BToken(ocean.web3, ocean.OCEAN_address) </br>
>>> assert OCEAN_token.balanceOf(alice_wallet.address) > 0, "need OCEAN" </br>
>>> pool = ocean.pool.create( </br>
...    token_address, </br>
...    data_token_amount=100.0, </br>
...    OCEAN_amount=10.0, </br>
...    from_wallet=alice_wallet </br>
... ) </br>
BPool.newBPool(). Begin. </br>
  pool_address = 0xb33b8DCEfe7A0b92Ea654d5908AB2c7Cb3050Ef1 </br>
BFactory.newBPool(). Done. </br>
>>> pool_address = pool.address </br>
>>> print(f"pool_address = '{pool_address}'") </br>
pool_address = '0xb33b8DCEfe7A0b92Ea654d5908AB2c7Cb3050Ef1' </br>
>>> from ocean_lib.common.agreements.service_types import ServiceTypes </br>
>>> asset = ocean.assets.resolve(did) </br>
>>> service1 = asset.get_service(ServiceTypes.ASSET_ACCESS) </br>
>>> pool = ocean.pool.get(ocean.web3, pool_address) </br>
>>> OCEAN_address = ocean.OCEAN_address </br>
>>> price_in_OCEAN = ocean.pool.calcInGivenOut( </br>
...     pool_address, OCEAN_address, token_address, token_out_amount=1.0) </br>
>>> print(f"Price of 1 datatoken is {price_in_OCEAN} OCEAN") </br>
Price of 1 datatoken is 0.9611175814077335 OCEAN </br>
>>> bob_wallet = Wallet(ocean.web3, private_key=os.getenv('TEST_PRIVATE_KEY2')) </br>
>>> print(f"bob_wallet.address = '{bob_wallet.address}'") </br>
bob_wallet.address = '0x33A4622B82D4c04a53e170c638B944ce27cffce3' </br>
>>> assert ocean.web3.eth.get_balance(bob_wallet.address) > 0, "need ganache ETH" </br>
>>> assert OCEAN_token.balanceOf(bob_wallet.address) > 0, "need ganache OCEAN" </br>
>>> data_token = ocean.get_data_token(token_address) </br>
>>> ocean.pool.buy_data_tokens( </br>
...     pool_address, </br>
...     amount=1.0, </br>
...     max_OCEAN_amount=10.0 </br>
...     from_wallet=bob_wallet </br>
... ) </br>
'0x8094ec768c61426a7d8d1b9646f859ed2ec9a94323f4385ddcb948d06180f87e' </br>
>>> print(f"Bob has {data_token.token_balance(bob_wallet.address)} datatokens.") </br>
Bob has 1.0 datatokens. </br>
>>> assert data_token.balanceOf(bob_wallet.address) >= 1.0, "Bob didn't get 1.0 datatokens" </br>
>>> fee_receiver = None  </br>
>>> from ocean_lib.common.agreements.service_types import ServiceTypes </br>
>>> asset = ocean.assets.resolve(did) </br>
>>> service = asset.get_service(ServiceTypes.ASSET_ACCESS) </br>
>>> quote = ocean.assets.order(asset.did, bob_wallet.address, service_index=service.index) </br>
>>> order_tx_id = ocean.assets.pay_for_service( </br>
...     ocean.web3, </br>
...     quote.amount, quote.data_token_address, asset.did, service.index, fee_receiver, bob_wallet, None) </br>
>>> print(f"order_tx_id = '{order_tx_id}'") </br>
order_tx_id = '0xd2deefc06d24ed437fed5cbacb979ff82e23f00126b97c3702319b84697622a7' </br>
>>> file_path = ocean.assets.download( </br>
...     asset.did, </br>
...     service.index, </br>
...     bob_wallet, </br>
...     order_tx_id, </br>
...     destination='./' </br>
... ) </br>
signing message with nonce 19: did:op:08787cC8dB325A9b84F60E99529C8000b4ef913c, account=0x33A4622B82D4c04a53e170c638B944ce27cffce3 </br>
>>> print(f"file_path = '{file_path}'") </br>
file_path = '/home/maria/Desktop/test3/datafile.0x08787cC8dB325A9b84F60E99529C8000b4ef913c.3' </br>
>>> </br>
</p>
</details>

**Comment by calina-c** [comment]
> This PR is missing the changes from #421. Please cherry-pick [0fc8d9b](https://github.com/oceanprotocol/ocean.py/commit/0fc8d9b3eb6ce8f9726b43a808020ea2097b510f) from the main branch.

The commit is not cherry-picked, but reproduced instead. It seems it's more sane like this, and resolving the conflicts from the cherry pick delayed the development process a lot. So I advised Maria to start from scratch instead of cherry picking.

File: ocean_lib/test/test_example_config.py
Comment (calina-c) [reply ocean_lib/test/test_example_config.py:1]: oh I now realise that part is not implemented. @mariacarmina even though you started from main and we decided to integrate the changes manually, step by step, instead of cherry picking and fixing conflicts, we still need to integrate the changes to get_web3_connection_provider.

**Comment by DMats** [comment]
> > This PR is missing the changes from #421. Please cherry-pick [0fc8d9b](https://github.com/oceanprotocol/ocean.py/commit/0fc8d9b3eb6ce8f9726b43a808020ea2097b510f) from the main branch.
> 
> The commit is not cherry-picked, but reproduced instead. It seems it's more sane like this, and resolving the conflicts from the cherry pick delayed the development process a lot. So I advised Maria to start from scratch instead of cherry picking.

I'd still like to see the contents of PR #421 merged into `main`, preferably added to this PR, regardless of how it's done. In particular I'd like to see:

1. Simplify get_web3_connection_provider
2. Removal of the infura related envvars, utils, and unit tests

**Comment by mariacarmina** [comment]
Since I added the changes from #421, I noticed that in compute-flow README, there is a config with the network `rinkeby`. Should I create a separate issue which points out the fact that we should include a part where the user needs to export the `rinkeby` network URL and then retrieves it from  `os` in the python code?

File: ocean_lib/test/test_example_config.py
Comment (mariacarmina) [reply ocean_lib/test/test_example_config.py:1]: I will use the public RPCs for the unit tests.

File: ocean_lib/example_config.py
Comment (mariacarmina) [reply ocean_lib/example_config.py:73]: I used the parameter `network_url` for `get_config-helper_network`, because I needed to add the network URL in the config dict, but I think I will add in the `get_config` function using `dict.update()` and implement your idea instead.

File: ocean_lib/test/test_example_config.py
Comment (mariacarmina) [reply ocean_lib/test/test_example_config.py:1]: It only worked for Polygon. Rinkeby, ropsten and mainnet are infura nodes which provide by default an infura URL that contains the secret key (saw on MetaMask).

**Comment by calina-c** [comment]
> Since I added the changes from #421, I noticed that in compute-flow README, there is a config with the network `rinkeby`. Should I create a separate issue which points out the fact that we should include a part where the user needs to export the `rinkeby` network URL and then retrieves it from `os` in the python code?

It's simple: you need to modify the readme to say that the user needs a NETWORK_URL, that will look something like 'http://rinkeby..../infura_id' etc. And paste it inside.

**Comment by mariacarmina** [comment]
> > Since I added the changes from #421, I noticed that in compute-flow README, there is a config with the network `rinkeby`. Should I create a separate issue which points out the fact that we should include a part where the user needs to export the `rinkeby` network URL and then retrieves it from `os` in the python code?
> 
> It's simple: you need to modify the readme to say that the user needs a NETWORK_URL, that will look something like 'http://rinkeby..../infura_id' etc. and paste it inside. Don't provide the rinkeby URL in the compute flow readme.

Ok, so need to export the URL as an envvar then. Thank you @calina-c

**Comment by trentmc** [comment]
Comments:

1. After the changes, does it respect this rule? "We can set any config parameter (a) via an envvar, or (b) via a config file. Envvar values override config file values." (as documented in [READMEs/parameters.md](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/parameters.md))
  - If no, we should either update change the PR to respect the rule, or update parameters.md with the updated rule

2. Redundant envvars.

> Changes proposed in this PR:
>  added `OCEAN_NETWORK_URL` env var;

There's _already_ a `NETWORK_URL` envvar.  `OCEAN_NETWORK_URL` would be redundant. Please resolve.

3. Update parameters.md

The two comments above point to things out of sync with how parameters.md describes things. It's likely that parameters.md itself needs updating, at the very least. Probably then other READMEs too!

I will probably have more comments once I git pull this and review in more detail myself.

**Comment by mariacarmina** [comment]
> Comments:
> 
> **A. Config param rules**
> 
> After the changes, does it respect this rule? "We can set any config parameter (a) via an envvar, or (b) via a config file. Envvar values override config file values." (as documented in [READMEs/parameters.md](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/parameters.md))
> 
> * If no, we should either update change the PR to respect the rule, or update parameters.md with the updated rule
> 
> **B. Redundant envvars.**
> 
> > Changes proposed in this PR:
> > added `OCEAN_NETWORK_URL` env var;
> 
> There's _already_ a `NETWORK_URL` envvar. `OCEAN_NETWORK_URL` would be redundant. Please resolve.
> 
> **C. parameters.md state**
> 
> The two comments above point to things out of sync with how parameters.md describes things. It's likely that parameters.md itself needs updating, at the very least. Probably then other READMEs too!
> 
> I will probably have more comments once I git pull this and review in more detail myself.

Regarding the redundant envvars, I have also discussed with @DMats about adding an envvar with the `OCEAN` prefix. I didn't see any usages of `NETWORK_URL` in the ocean.py codebase, only in READMEs. Should I switch using `NETWORK_URL` instead of `OCEAN_NETWORK_URL` or should I update the READMEs?

**Comment by trentmc** [comment]
> Regarding the redundant envvars, I have also discussed with @DMats about adding an envvar with the OCEAN prefix. I didn't see any usages of NETWORK_URL in the ocean.py codebase, only in READMEs. 

You missed the critical one in `ocean_lib/config.py` with `NAME_NETWORK_URL: ["NETWORK_URL"`. All it takes is one critical line.

> Should I switch using NETWORK_URL instead of OCEAN_NETWORK_URL or should I update the READMEs?

**Summary:** keep `NETWORK_URL` for _this_ issue.

**Details:**

Each PR should be as small as possible to meet its issue. It makes it easier to review, and for anyone looking back at the history of the code.

If you're trying to solve a different issue, then create a different issue and do the PR there.

The scope of this PR is already >> the original issue. 

Therefore for this issue, I recommend: 
- creating a new issue now. E.g. "Rename NETWORK_URL -> OCEAN_NETWORK_URL` for consistency with ocean.js". Make _this_ issue a blocker to completing that issue
- ensure that _this_ uses NETWORK_URL, to keep down changes to scope

**Comment by mariacarmina** [comment]
> > Regarding the redundant envvars, I have also discussed with @DMats about adding an envvar with the OCEAN prefix. I didn't see any usages of NETWORK_URL in the ocean.py codebase, only in READMEs.
> 
> You missed the critical one in `ocean_lib/config.py` with `NAME_NETWORK_URL: ["NETWORK_URL.."`. All it takes is one critical line.
> Ok, so that one is binded with the actual envvar. Got it, thanks!
> > Should I switch using NETWORK_URL instead of OCEAN_NETWORK_URL or should I update the READMEs?
> 
> **Summary:** keep `NETWORK_URL` for _this_ issue.
> 
> **Details:**
> 
> Each PR should be as small as possible to meet its issue. It makes it easier to review, and for anyone looking back at the history of the code.
> 
> If you're trying to solve a different issue, then create a different issue and do the PR there.
> 
> The scope of this PR is already >> the original issue.
> 
> Therefore for this issue, I recommend:
> 
> * creating a new issue now. E.g. "Rename NETWORK_URL -> OCEAN_NETWORK_URL` for consistency with ocean.js". Make _this_ issue a blocker to completing that issue
> * ensure that _this_ uses NETWORK_URL, to keep down changes to scope
> Sounds good.

File: READMEs/compute-flow.md
Comment (mariacarmina) [inline READMEs/compute-flow.md:86]: Should I keep the same logic here and replace only the envvar, @trentmc?

Comment (trentmc) [reply READMEs/compute-flow.md:86]: > Should I keep the same logic here and replace only the envvar, @trentmc?

Yes, of course

**Comment by trentmc** [comment]
@mariacarmina please ensure that you've addressed all three comments (A, B, C) above. So far the discussion has only really been around (B).

**Comment by mariacarmina** [comment]
> @mariacarmina please ensure that you've addressed all three comments (A, B, C) above. So far the discussion has only really been around (B).


For A, I assume that the flow is reproduced, because I have some unit tests in `test_example_config` which use different RPCs (different from the ganache one that is present in the config file) and they passed. For C, I have also noticed that `parameters.md` needs refactor in some places (in the code as well). Maybe we should create an issue for that.

**Comment by trentmc** [comment]
> For A, I assume that the flow is reproduced, because I have some unit tests in test_example_config which use different RPCs (different from the ganache one that is present in the config file) and they passed. 

Do your changes still respect the rule "We can set any config parameter (a) via an envvar, or (b) via a config file. Envvar values override config file values."  ? (This is what A is asking)


> For C, I have also noticed that parameters.md needs refactor in some places (in the code as well). Maybe we should create an issue for that.

If your changes break parameters.md, then you need to change them as part of this PR. No PR should ever be merged if it's still causing breakage.

**Comment by mariacarmina** [comment]
**Summary**

(A) 
My changes still respect the rule from the `parameters.md`. It's tested through the `test_example_config.py` unit tests, where the network URL differs from the config.ini network URL.
(B) 
I reverted to `NETWORK_URL` [here](https://github.com/oceanprotocol/ocean.py/pull/435/commits/e01c6b7f87b947e1d8c4785498430ccfddf4c28d). Created #442 to replace NETWORK_URL with OCEAN_NETWORK_URL for consistency.
(C) 
My changes do not break `parameters.md`. I saw one issue before my PR and reported it in #449.

**Comment by calina-c** [comment]
This was approved by Trent, but he is too busy to comment right now. I'll let the tests run and then merge it.

**Pull request merged**
